crates/sui-rpc-api/proto/sui/rpc/v2alpha/archival_service.proto

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+syntax = "proto3";
+
+package sui.rpc.v2alpha;
+
+import "google/protobuf/field_mask.proto";
+import "sui/rpc/v2alpha/checkpoint.proto";
+import "sui/rpc/v2alpha/executed_transaction.proto";
+import "sui/rpc/v2alpha/object.proto";
+
+// Alternitive names:
+// - HistoryService
+// - LedgerService
+// - HistoricalService
+// - HistoricalDataService
+service ArchivalService {
+  rpc GetObject(GetObjectRequest) returns (GetObjectResponse);
+  rpc BatchGetObjects(BatchGetObjectsRequest) returns (BatchGetObjectsResponse);
+
+  rpc GetTransaction(GetTransactionRequest) returns (GetTransactionResponse);
+  rpc BatchGetTransactions(BatchGetTransactionsRequest) returns (BatchGetTransactionsResponse);
+
+  rpc GetCheckpoint(GetCheckpointRequest) returns (GetCheckpointResponse);
+  rpc BatchGetCheckpoints(BatchGetCheckpointsRequest) returns (BatchGetCheckpointsResponse);
+}
+
+message GetObjectRequest {
+  // Required. The `ObjectId` of the requested object.
+  optional string object_id = 1;
+
+  // Request a specific version of the object.
+  // If no version is specified, and the object is live, then the latest
+  // version of the object is returned.
+  optional uint64 version = 2;
+
+  // Mask specifying which fields to read.
+  // If no mask is specified, defaults to `object_id,version,digest`.
+  optional google.protobuf.FieldMask read_mask = 4;
+}
+
+message GetObjectResponse {
+  optional Object object = 1;
+}
+
+message BatchGetObjectsRequest {
+  repeated GetObjectRequest requests = 1;
+
+  // Mask specifying which fields to read.
+  // If no mask is specified, defaults to `object_id,version,digest`.
+  optional google.protobuf.FieldMask read_mask = 4;
```

Comment:
Are the field indices intentionally gappy here, because we're already using this schema in production and this is an evolution of it, or is it copypasta?

Reply:
There were a few messages I defined that intentionally were gappy this one is probably copypasta though. Field number is really only important for micro optimizations in that 1-15 are encoded as a single byte where as larger numbers require more bytes on the wire.